src/json/json.cc

```diff
@@ -0,0 +1,3232 @@
+                            INT_MAX, nullptr, Config_GetHashTableGrowFactor, Config_SetHashTableGrowFactor)
+    REGISTER_NUMERIC_CONFIG(ctx, "hash-table-shrink-factor", 50, VALKEYMODULE_CONFIG_DEFAULT, 0,
+                            INT_MAX, nullptr, Config_GetHashTableShrinkFactor, Config_SetHashTableShrinkFactor)
+    REGISTER_NUMERIC_CONFIG(ctx, "hash-table-min-load-factor", 25, VALKEYMODULE_CONFIG_DEFAULT, 0,
+                            INT_MAX, nullptr, Config_GetHashTableMinLoadFactor, Config_SetHashTableMinLoadFactor)
+    REGISTER_NUMERIC_CONFIG(ctx, "hash-table-max-load-factor", 85, VALKEYMODULE_CONFIG_DEFAULT, 0,
+                            INT_MAX, nullptr, Config_GetHashTableMaxLoadFactor, Config_SetHashTableMaxLoadFactor)
+    REGISTER_NUMERIC_CONFIG(ctx, "hash-table-min-size", 64, VALKEYMODULE_CONFIG_DEFAULT, 0, INT_MAX,
+                            nullptr, Config_GetHashTableMinSize, Config_SetHashTableMinSize)
+
+    ValkeyModule_LoadConfigs(ctx);
+    return VALKEYMODULE_OK;
+}
+
+/*
+ * Install stub datatype callback for aux_load.
+ */
+bool install_stub(ValkeyModuleCtx *ctx,
+                  const char *type_name,
+                  int encver,
+                  int (*aux_load)(ValkeyModuleIO *, int encver, int when)) {
+    ValkeyModuleTypeMethods type_methods;
+    memset(&type_methods, 0, sizeof(ValkeyModuleTypeMethods));
+    type_methods.version = VALKEYMODULE_TYPE_METHOD_VERSION;
+    type_methods.aux_load = aux_load;
+    if (ValkeyModule_CreateDataType(ctx, type_name, encver, &type_methods) == nullptr) {
+        ValkeyModule_Log(ctx, "warning", "Failed to create data type %s", type_name);
+        return false;
+    }
+    ValkeyModule_Log(ctx, "debug", "Successfully installed stub data type %s", type_name);
+    return true;
+}
+
+/*
+ * Check a string value, fail if the expected value isn't present.
+ */
+bool checkString(ValkeyModuleIO *ctx, const char *value, const char *caller) {
+    size_t str_len;
+    std::unique_ptr<char> str(ValkeyModule_LoadStringBuffer(ctx, &str_len));
+    if (strncmp(value, str.get(), str_len)) {
+        ValkeyModule_Log(nullptr, "warning", "%s: Unexpected value in RDB. Expected %s Received %s",
+                        caller, value, str.get());
+        return false;
+    }
+    return true;
+}
+
+/*
+ * Check an integer value, fail
+ */
+bool checkInt(ValkeyModuleIO *ctx, uint64_t value, const char *caller) {
+    uint64_t val = ValkeyModule_LoadUnsigned(ctx);
+    if (value != val) {
+        ValkeyModule_Log(nullptr, "warning", "%s: Unexpected value in RDB Expected: %lx Received: %lx",
+                        caller, value, val);
+        return false;
+    }
+    return true;
+}
+
+/*
+ * Check the encoding version, For unsupported versions we ALWAYS put out a message in the log
+ * but we only fail the RDB load if the config tells us to do it.
+ */
+bool checkVersion(const char *type_name, int encver, int expected_encver) {
```

Comment:
I suppose I don't get what this is for. What does this do for compatibility?

Reply:
The function is invoked when loading a RDB produced by RedisJSON. Their RDB could include a newer version of a module type. This function decides whether or not to reject the RDB.

Reply:
Default value of enforce_rdb_version_check is false. So, the check will pass. In our internal RDB compatibility tests, which isn't part of the PR, we set enforce_rdb_version_check to true so that we will be alerted when ReJSON introduced some new aux types. In retrospect, I think the aux type version checking is an overkill. We should just read and ignore these aux types.

Reply:
https://github.com/valkey-io/valkeyJSON/issues/12